R/method-correlation.R

```diff
@@ -0,0 +1,100 @@
+#' Correlation
+#'
+#' This function returns correlation coefficients for all abundance variables vs all continuous sample metadata variables.
+#' Soon this function will be expanded to take multiple abundance datasets, and to run all-to-all correlations
+#' for abundance variables.
+#' 
+#' @param data AbundanceData object
+#' @param method string defining the type of correlation to run. The currently supported values are 'spearman' and 'pearson'
+#' @param verbose boolean indicating if timed logging is desired
+#' @return ComputeResult object
+#' @import veupathUtils
+#' @import data.table
+#' @useDynLib microbiomeComputations
+#' @export
+setGeneric("correlation",
+  function(data, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) standardGeneric("correlation"),
+  signature = c("data")
+)
+
+#'@export
+setMethod("correlation", signature("AbundanceData"), function(data, method = c('spearman','pearson'), verbose = c(TRUE, FALSE)) {
+    df <- data@data
+    recordIdColumn <- data@recordIdColumn
+    naToZero <- data@imputeZero
+    ancestorIdColumns <- data@ancestorIdColumns
+    allIdColumns <- c(recordIdColumn, ancestorIdColumns)
+    sampleMetadata <- data@sampleMetadata
+
+    ## Initialize and check inputs
+    method <- veupathUtils::matchArg(method)
+    verbose <- veupathUtils::matchArg(verbose)
+
+    # Check that incoming df meets requirements
+    if (!'data.table' %in% class(df)) {
+      data.table::setDT(df)
+    }
+
+    # Check that incoming metadata meets requirements
+    if (!'data.table' %in% class(sampleMetadata)) {
+      data.table::setDT(sampleMetadata)
+    }
+
+    computeMessage <- ''
+    veupathUtils::logWithTime(paste("Received df table with", nrow(df), "samples and", (ncol(df)-1), "taxa."), verbose)
+
+    # Replace NA values with 0
+    if (naToZero) {
+      veupathUtils::setNaToZero(df)
```

Comment:
im starting to wonder if this should be a method of AbundanceData rather than called by each of its methods individually

Reply:
i could see the different methods having opinions about setting na to zero or not. Like maybe one correlation method wants to use it but another doesn't. I'll learn more when we actually add the more interesting correlation methods

Reply:
imputing zeroes seems to be about appropriately modeling the data, not about the compute? not sure i understand

Reply:
so we probably could have the class do it automatically, but im not sure thats what we want. i could see keeping track of the original/ unmodified data being worthwhile. i think what id prefer is an accessor fxn for the AbundanceData class. we should probably be writing those anyhow, but this maybe increases the pressure there.. the idea i think would be to write a method for AbundanceData like `getDataTable` or something, which by default imputes zeroes if necessary, but can be overridden if desired.